README.md
# Biblepay Ubuntu Package Builder

This project is used to build Ubuntu packages for Biblepay and upload them to Launchpad/PPA.

(PPA is an repository for deb-packages that is hosted by ubuntus launchpad)

## Requirements

You need a linux system (ubuntu or debian is not required) with: 
* bash  (indeed)
* docker

Everything else is installed inside a docker container.

## Preparation

Before you can start with building the package, you must create an PPA on launchpa

**1. Create a GnuPG Key**
Complete howto: https://help.ubuntu.com/community/GnuPrivacyGuardHowto

* Install gnupg
* Run "gpg --gen-key"
* Choose "RSA and RSA (default)"
* Use your real e-mail adress!
* Choose a secure key password
* You will need the Fingerprint, copy it
* Upload the public key with "gpg --send-keys --keyserver keyserver.ubuntu.com [KEY-ID]" (KEY-ID example: "D8FC66D2")
* Do not forget to backup your keys!

**2. Register with launchpad**
* Go to https://launchpad.net
* Register yourself
* Add Fingerprint (Click on your username at the top-right corner, then "Change email settings"). You find the     fingerprint in the output 
* Launchpad will send you an email to ensure that you own the private key of the fingerprint. Copy the Message (starts with "-----BEGIN PGP MESSAGE-----" and ends with "-----END PGP MESSAGE-----" and save it to a file. Decrypt file with "gpg --output [OUTPUTFILENAME] --decrypt [ENCRYPTED FILE]"
* The decryped message will contain a link, open it to verify your ownership of the key 

**3. Register a team at launchpad**
This step is only required if there is not already a team existing. If that is the case, request membership. You can also skip this, if you want to use your own personal ppa.

* Register a team at https://launchpad.net/people/+newteam

**4. Create a PPA**

* Go to your Team page (or personal page) in Launchpad and click "Create a new PPA"
* If this is your first time with Launchpad, you might want to create a test PPA first
* Choose usefull names. Remember that you can not change them later
* IMPORTANT: The biblepay Client required an old Version of the berkeley db that is part of the bitcoin ppa repository. Go to your PPA page and click "Edit PPA dependencies". Search for "bitcoin" and add the "bitcoin/ubuntu/bitcoin" repository as dependency.

## Configuration

All you need todo is copy the file "ppa.conf.USE" as "ppa.conf" and update it:

```
[my-ppa]
fqdn = ppa.launchpad.net
method = ftp
incoming = ~<your_user-or-team-launchpad_id>/ubuntu/<ppa_name>/
login = anonymous
allow_unsigned_uploads = 0
```
Change <your_user-or-team-launchpad_id> and <ppa_name> to your own values.
(This file will be used as dput.cf and copied to /root/ in the docker image)

## Build for ppa

Before you can build a new version, you must update the file "debian/changelog" and add a new entry to the file. Important: Newer entries are at the TOP!

Then this is done, simply call **./ppa_build.sh**. The script will ask you some questions, answer them and the building will start.

**Important:** Only the source code packages are build locally, the real binary packages are build on launchpad. You can see the build status on your launchpad team site. It might take 1-2 hours to build the packages.

# Build binaries locally 

Call **./local_build.sh** and follow the instructions. You will find the packages in the newly created folder "packages"


## Notes

* The scripts will build packages for all ubuntu releases given in "ubuntu_releases". The file MUST end with an empty line!


## Possible problems

* If the build fails, ensure that you added the Dependencies (bitcoint ppa) to your ppa repository!
* If you regenerate the same version again and upload it to PPA, launchpad will refect it. You might try this if you forgot to add the dependencies, but you must change the ubuntu version number (the one after [release]) in the changelog.


debian/changelog
biblepay (1.1.1.9-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 06 Apr 2018 08:00:00 +0100

biblepay (1.1.1.8b-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 03 Apr 2018 10:50:00 +0100

biblepay (1.1.1.3-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 25 Mar 2018 10:50:00 +0100

biblepay (1.1.1.2-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 23 Mar 2018 20:50:00 +0100

biblepay (1.1.1.1f-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 20 Mar 2018 18:48:00 +0100

biblepay (1.1.1.1c-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 16 Mar 2018 23:04:00 +0100

biblepay (1.1.1.1b-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 13 Mar 2018 21:20:00 +0100

biblepay (1.1.1.0-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 11 Mar 2018 14:50:00 +0100

biblepay (1.1.0.9b-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 10 Mar 2018 18:28:00 +0100

biblepay (1.1.0.8-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 10 Mar 2018 10:10:00 +0100

biblepay (1.1.0.6-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 08 Mar 2018 19:37:00 +0100

biblepay (1.1.0.2c-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 06 Mar 2018 18:37:00 +0100

biblepay (1.0.9.9-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 05 Mar 2018 19:30:00 +0100

biblepay (1.0.9.7-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 04 Mar 2018 18:00:00 +0100

biblepay (1.0.8.4-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 22 Jan 2018 20:33:00 +0100

biblepay (1.0.8.3c-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 20 Jan 2018 11:26:00 +0100

biblepay (1.0.8.2-[release]1) [release]; urgency=low
  * New upstream release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 19 Jan 2018 09:02:04 +0100

biblepay (1.0.8.1b-[release]1) [release]; urgency=low
  * First debian release
 -- Lichtsucher <lichtsucher@maxolero.net>  Thu, 17 Jan 2018 17:16:04 +0100

ppa_build.sh
#!/usr/bin/env bash

# get the absolute path to this script. Required to mount the "packages" directory
BASEDIR="$( cd "$(dirname "$0")" ; pwd -P )"



# '" ensure that the changelog is ok

echo ""
echo "Before you start, ensure that you updated debian/changelog! Here are the first 20 lines:"
echo ""

head -n 20 debian/changelog

while true; do
    echo ""
    echo ""
    read -p "Is the changelog correct? (y/n)" yn
    case $yn in
        [Yy]* ) break;;
        [Nn]* ) exit;;
        * ) echo "Please answer [y]es or [n]o.";;
    esac
done



# request the version number

echo "Please enter the Biblepay version string off the new release (like: 1.8.0.1). Do NOT add the DEB-Subversion (-1, like 1.8.0.1-1)"

read VERSIONSTR

echo "Entered $VERSIONSTR"


# final check
while true; do
    echo ""
    echo ""
    read -p "The script will now start the docker image, pull the current HEAD and build it as biblepay version $VERSIONSTR. Ok? (y/n)" yn
    case $yn in
        [Yy]* ) break;;
        [Nn]* ) exit;;
        * ) echo "Please answer [y]es or [n]o.";;
    esac
done


# output directory for the deb-files
mkdir -p $BASEDIR/packages/ppa_orig


# read the ubuntu version from the file and put it into an array
# we need to have an own step here, or the loop would stop the tty, which we need for docker
while read release; do
  releases+=("$release")
done < ubuntu_releases

# build a version of the package for every given ubuntu release
for line in "${releases[@]}"; do
  # nice trick that splits the string "xenial|16.04" into "xenial" and "16.04"
  RELEASENAME=${line%|*}
  RELEASEVERSION=${line#*|}

  echo "Build package for $RELEASENAME / $RELEASEVERSION"

  # create a temp folder and place a release-specific Dockerfile there
  sed 's/\[releaseversion\]/16.04/g' Dockerfile.tpl  > Dockerfile

  # build the docker image, won't execute any code
  docker build -t "debpackage-$RELEASENAME" .

  # start the docker image with some required volumes and starts the build process itselfs
  # we do one buildprocess per ubuntu version
  docker run -it -e RELEASENAME="$RELEASENAME" -e RELEASEVERSION="$RELEASEVERSION" -e VERSIONSTR="$VERSIONSTR" -v $BASEDIR/packages:/packages -v $HOME/.gnupg:/root/.gnupg -v $BASEDIR/debian:/debian "debpackage-$RELEASENAME" /bin/sh /code/run_ppa_build.sh

  # cleanup
  rm -f Dockerfile
done







local_build.sh
#!/usr/bin/env bash

# get the absolute path to this script. Required to mount the "packages" directory
BASEDIR="$( cd "$(dirname "$0")" ; pwd -P )"



# ensure that the changelog is ok

echo ""
echo "Before you start, ensure that you updated debian/changelog! Here are the first 20 lines:"
echo ""

head -n 20 debian/changelog

while true; do
    echo ""
    echo ""
    read -p "Is the changelog correct? (y/n)" yn
    case $yn in
        [Yy]* ) break;;
        [Nn]* ) exit;;
        * ) echo "Please answer [y]es or [n]o.";;
    esac
done



# request the version number

echo "Please enter the Biblepay version string off the new release (like: 1.8.0.1). Do NOT add the DEB-Subversion (-1, like 1.8.0.1-1)"

read VERSIONSTR

echo "Entered $VERSIONSTR"


# final check
while true; do
    echo ""
    echo ""
    read -p "The script will now start the docker image, pull the current HEAD and build it as biblepay version $VERSIONSTR. Ok? (y/n)" yn
    case $yn in
        [Yy]* ) break;;
        [Nn]* ) exit;;
        * ) echo "Please answer [y]es or [n]o.";;
    esac
done


# output directory for the deb-files
# The ppa_orig directory saves the old ppa files, so that we do not change them between uploads
mkdir -p $BASEDIR/packages/


# read the ubuntu version from the file and put it into an array
# we need to have an own step here, or the loop would stop the tty, which we need for docker
while read release; do
  releases+=("$release")
done < ubuntu_releases

# build a version of the package for every given ubuntu release
for line in "${releases[@]}"; do
  # nice trick that splits the string "xenial|16.04" into "xenial" and "16.04"
  RELEASENAME=${line%|*}
  RELEASEVERSION=${line#*|}

  echo "Build package for $RELEASENAME / $RELEASEVERSION"

  # create a temp folder and place a release-specific Dockerfile there
  sed 's/\[releaseversion\]/16.04/g' Dockerfile.tpl  > Dockerfile

  # build the docker image, won't execute any code
  docker build -t "debpackage-$RELEASENAME" .

  # start the docker image with some required volumes and starts the build process itselfs
  # we do one buildprocess per ubuntu version
  docker run -it -e RELEASENAME="$RELEASENAME" -e RELEASEVERSION="$RELEASEVERSION" -e VERSIONSTR="$VERSIONSTR" -v $BASEDIR/packages:/packages -v $HOME/.gnupg:/root/.gnupg -v $BASEDIR/debian:/debian "debpackage-$RELEASENAME" /bin/sh /code/run_local_build.sh

  # cleanup
  rm -f Dockerfile
done







helper/run_ppa_build.sh
# we remove all old files that might exist
rm -rf /code/*

# launchpad hates it when the same *.orig"-File was created two times
# with slightly differen time stamps. So we only create one, if that is
# really necessary

if [ ! -f /packages/ppa_orig/biblepay_$VERSIONSTR.orig.tar.bz2 ]; then
  # file not found

  echo "No existing orig-file found, clone from git"

  # we clone of the biblepay sourcecode here
  git clone https://github.com/biblepay/biblepay biblepay
  #git clone https://github.com/Lichtsucher/biblepay

  # we build the "orignal source" from the cloned git repository (excluding the git files)
  cd /code/biblepay
  tar cfj ../biblepay_$VERSIONSTR.orig.tar.bz2 . --exclude-vcs
  cp ../biblepay_$VERSIONSTR.orig.tar.bz2 /packages/ppa_orig/
else
  # found

  echo "Existing orig-file found. Use it"

  cp /packages/ppa_orig/biblepay_$VERSIONSTR.orig.tar.bz2 /code/

  mkdir biblepay
  cd /code/biblepay
  tar -jxf ../biblepay_$VERSIONSTR.orig.tar.bz2
fi

# the debian folder must be in debian/ not contrib/debian, so we must move it
cp -rf contrib/debian .

# here we overwrite or add files to the debian folder
cp -rf /debian/* ./debian/

# update the changelog and replace [release] with $RELEASENAME
mv debian/changelog debian/changelog_org
sed "s/\[release\]/$RELEASENAME/g" debian/changelog_org > debian/changelog

# ensure that important files are executable
chmod 777 share/genbuild.sh
chmod 777 autogen.sh 

# create and sign the required debian source files
debuild -S -sa

# after the source files are finished, we upload them to ppa
# the name "biblepay" in dput" is a reference to the section [biblepay] in the ppa.conf
cd /code/
dput biblepay *_source.changes

# we also copy all created files to our packages directory
cp /code/* /packages/


helper/run_local_build.sh
# we remove all old files that might exist
rm -rf /code/*

# we clone of the biblepay sourcecode here
git clone https://github.com/biblepay/biblepay biblepay
#git clone https://github.com/Lichtsucher/biblepay

cd /code/biblepay

# we build the "orignal source" from the cloned git repository (excluding the git files)
tar cfj ../biblepay_$VERSIONSTR.orig.tar.bz2 . --exclude-vcs

# the debian folder must be in debian/ not contrib/debian, so we must move it
cp -rf contrib/debian .

# here we overwrite or add files to the debian folder
cp -rf /debian/* ./debian/

# update the changelog and replace [release] with $RELEASENAME
mv debian/changelog debian/changelog_org
sed "s/\[release\]/$RELEASENAME/g" debian/changelog_org > debian/changelog

# ensure that important files are executable
chmod 777 share/genbuild.sh
chmod 777 autogen.sh 

# create and sign the required debian source files
debuild

# we now copy the finished files to the packages volume that is a directory of the host system
cp /code/*.deb /packages/


ppa_nightly_build.sh
#!/usr/bin/env bash

# get the absolute path to this script. Required to mount the "packages" directory
BASEDIR="$( cd "$(dirname "$0")" ; pwd -P )"

# And dates as version numbers '"
CURRENT_DATE=`LC_ALL=en_US date '+%Y%m%d%H%M%S'`
CURRENT_DATE_NICE=`LC_ALL=en_US date '+%a, %d %b %Y %H:%M:00 +0100'`


# First, we update the repository and get the current (and last) version
# You must download this repository by yourself before you can use the script
cd bbp_nightly_repository
git pull origin master
cd ..

CURRENT_VERSION=`cd bbp_nightly_repository && git rev-parse HEAD`
echo $CURRENT_VERSION

touch last_nightly_commit.data
LAST_VERSION=`cat last_nightly_commit.data`

# Is the current newest commit different fromt last one?
# If not, then exit
if [ "$CURRENT_VERSION" == "$LAST_VERSION" ]; then
   echo "No change found. Exit"
   exit
fi


# write the changelog
# We save the old one
cp debian/changelog debian/changelog.save

echo "biblepay ($CURRENT_DATE-[release]1) [release]; urgency=low" > debian/changelog
echo "  * New upstream release" >> debian/changelog
echo -n " -- Lichtsucher <lichtsucher@maxolero.net>  $CURRENT_DATE_NICE" >> debian/changelog

# output directory for the deb-files
/bin/mkdir -p $BASEDIR/packages/ppa_orig

# read the ubuntu version from the file and put it into an array
# we need to have an own step here, or the loop would stop the tty, which we need for docker
while read release; do
  releases+=("$release")
done < ubuntu_releases

# build a version of the package for every given ubuntu release
for line in "${releases[@]}"; do
  # nice trick that splits the string "xenial|16.04" into "xenial" and "16.04"
  RELEASENAME=${line%|*}
  RELEASEVERSION=${line#*|}

  echo "Build package for $RELEASENAME / $RELEASEVERSION"

  # create a temp folder and place a release-specific Dockerfile there
  sed 's/\[releaseversion\]/16.04/g' Dockerfile.tpl  > Dockerfile

  # build the docker image, won't execute any code
  docker build -t "debpackage-$RELEASENAME" .

  # start the docker image with some required volumes and starts the build process itselfs
  # we do one buildprocess per ubuntu version
  docker run -it -e RELEASENAME="$RELEASENAME" -e RELEASEVERSION="$RELEASEVERSION" -e VERSIONSTR="$CURRENT_DATE" -v $BASEDIR/packages:/packages -v $HOME/.gnupg:/root/.gnupg -v $BASEDIR/debian:/debian "debpackage-$RELEASENAME" /bin/sh /code/run_ppa_build.sh

  # cleanup
  rm -f Dockerfile
done



# Cleanup and version saving

echo -n "$CURRENT_VERSION" > last_nightly_commit.data
cp debian/changelog.save debian/changelog

